Bticino dimmer

Hello !

I have somme dimmer in my house, and with actual Bticino binding we can't manadge them, only on and off...

For do that, some change needed in the code : 
https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.bticino/src/main/java/org/openhab/binding/bticino/internal/BticinoDevice.java#L141

And add some change like this : 
> case 1: {
>            if (OnOffType.ON.equals(command)) {
>                l_pr.addProperty("what", "1");
>            **} elseif (DimmerType) {**
>                            **whatvalue = searching percentage and divide by 10, round 1 (ie 0 to 10 is the result, nothing else)**
>                             **l_pr.addProperty("what", "whatvalue");**
>             } else {
>                l_pr.addProperty("what", "0");
>            }
>            break;

I'm sorry, i'm not so strong to code in java...

Thanks a lot,
e4.

Hello !

Increase/decrease work successfully !


```
23:07:54.608 [DEBUG] [bticino.internal.BticinoDevice:134  ] - Gateway [default], Command 'INCREASE' received for item Light_Salon_Dimm
23:07:54.612 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*32]
23:07:54.614 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:07:54.617 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : address, Value : 32
23:07:54.620 [DEBUG] [bticino.internal.BticinoDevice:147  ] - Light received INCREASE command.
23:07:54.623 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : what, Value : 30
23:07:54.638 [INFO ] [.m.f.queue.PriorityQueueThread:79   ] - OpenWebNet CMD [*1*30*32##]
23:07:54.710 [INFO ] [penwebnet.MonitorSessionThread:71   ] - Received OpenWebNet frame '*1*10*32##' now translate it to an event.
23:07:54.713 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*10*32]
23:07:54.715 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:07:54.716 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : what, Value : 10
23:07:54.718 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : where, Value : 32
23:07:54.719 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : virtual, Value : false
23:07:54.721 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageType, Value : Lighting
23:07:54.722 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageDescription, Value : Light Dimmer
23:07:54.724 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.class, Value : Light
23:07:54.727 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.name, Value : 1*32
23:07:54.729 [INFO ] [penwebnet.MonitorSessionThread:714  ] - Light Dimmer
23:07:54.730 [DEBUG] [bticino.internal.BticinoDevice:216  ] - Gateway [default], Bticino WHO [1], WHAT [10], WHERE [32]
23:07:54.732 [DEBUG] [bticino.internal.BticinoDevice:239  ] - Gateway [default], RECEIVED EVENT FOR SwitchItem [Light_Salon_Dimm], TRANSLATE TO OPENHAB BUS EVENT
23:07:54.746 [INFO ] [runtime.busevents             :22   ] - Light_Salon_Dimm received command INCREASE
23:07:54.748 [DEBUG] [bticino.internal.BticinoDevice:134  ] - Gateway [default], Command 'INCREASE' received for item Light_Salon_Dimm
23:07:54.751 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*32]
23:07:54.754 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:07:54.757 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : address, Value : 32
23:07:54.760 [DEBUG] [bticino.internal.BticinoDevice:147  ] - Light received INCREASE command.
23:07:54.764 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : what, Value : 30
23:07:54.950 [INFO ] [.m.f.queue.PriorityQueueThread:79   ] - OpenWebNet CMD [*1*30*32##]
23:07:55.030 [INFO ] [penwebnet.MonitorSessionThread:71   ] - Received OpenWebNet frame '*1*10*32##' now translate it to an event.
23:07:55.033 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*10*32]
23:07:55.034 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:07:55.036 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : what, Value : 10
23:07:55.037 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : where, Value : 32
23:07:55.038 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : virtual, Value : false
23:07:55.040 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageType, Value : Lighting
23:07:55.041 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageDescription, Value : Light Dimmer
23:07:55.043 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.class, Value : Light
23:07:55.045 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.name, Value : 1*32
23:07:55.047 [INFO ] [penwebnet.MonitorSessionThread:714  ] - Light Dimmer
23:07:55.048 [DEBUG] [bticino.internal.BticinoDevice:216  ] - Gateway [default], Bticino WHO [1], WHAT [10], WHERE [32]
```
or 

```23:09:21.813 [INFO ] [penwebnet.MonitorSessionThread:714  ] - Light Dimmer
23:09:21.815 [DEBUG] [bticino.internal.BticinoDevice:216  ] - Gateway [default], Bticino WHO [1], WHAT [7], WHERE [32]
23:09:21.816 [DEBUG] [bticino.internal.BticinoDevice:239  ] - Gateway [default], RECEIVED EVENT FOR SwitchItem [Light_Salon_Dimm], TRANSLATE TO OPENHAB BUS EVENT
23:09:22.031 [INFO ] [.m.f.queue.PriorityQueueThread:79   ] - OpenWebNet CMD [*1*31*32##]
23:09:22.139 [INFO ] [penwebnet.MonitorSessionThread:71   ] - Received OpenWebNet frame '*1*6*32##' now translate it to an event.
23:09:22.142 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*6*32]
23:09:22.144 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:09:22.145 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : what, Value : 6
23:09:22.146 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : where, Value : 32
23:09:22.147 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : virtual, Value : false
23:09:22.149 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageType, Value : Lighting
23:09:22.150 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : messageDescription, Value : Light Dimmer
23:09:22.152 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.class, Value : Light
23:09:22.154 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : object.name, Value : 1*32
23:09:22.155 [INFO ] [penwebnet.MonitorSessionThread:714  ] - Light Dimmer
23:09:22.156 [DEBUG] [bticino.internal.BticinoDevice:216  ] - Gateway [default], Bticino WHO [1], WHAT [6], WHERE [32]```

But when i use my Android apps, they don't work : 

```
23:11:53.772 [DEBUG] [o.i.r.i.resources.ItemResource:205  ] - Received HTTP POST request at 'items/Light_Salon_Dimm' with value '26'.
23:11:53.790 [INFO ] [runtime.busevents             :22   ] - Light_Salon_Dimm received command 26
23:11:53.795 [DEBUG] [bticino.internal.BticinoDevice:134  ] - Gateway [default], Command '26' received for item Light_Salon_Dimm
23:11:53.799 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*32]
23:11:53.802 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:11:53.805 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : address, Value : 32
23:11:53.808 [WARN ] [bticino.internal.BticinoDevice:177  ] - Received unknown command type for lighting: 'org.openhab.core.library.types.PercentType'
23:11:53.822 [ERROR] [vlaminck.openwebnet.OpenWebNet:94   ] - onCommand error : null
23:11:54.413 [DEBUG] [o.o.i.r.i.b.GeneralBroadcaster:46   ] - broadcaster 'org.openhab.io.rest.internal.broadcaster.GeneralBroadcaster$1@1b38868' is empty
23:11:54.413 [DEBUG] [o.o.i.r.i.b.GeneralBroadcaster:46   ] - broadcaster 'org.openhab.io.rest.internal.broadcaster.GeneralBroadcaster$1@1b38868' is empty
```

or 

```

23:12:54.679 [DEBUG] [o.i.r.i.resources.ItemResource:205  ] - Received HTTP POST request at 'items/Light_Salon_Dimm' with value '53'.
23:12:54.685 [INFO ] [runtime.busevents             :22   ] - Light_Salon_Dimm received command 53
23:12:54.688 [DEBUG] [bticino.internal.BticinoDevice:134  ] - Gateway [default], Command '53' received for item Light_Salon_Dimm
23:12:54.691 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*32]
23:12:54.694 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:12:54.697 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : address, Value : 32
23:12:54.700 [WARN ] [bticino.internal.BticinoDevice:177  ] - Received unknown command type for lighting: 'org.openhab.core.library.types.PercentType'
23:12:54.703 [ERROR] [vlaminck.openwebnet.OpenWebNet:94   ] - onCommand error : null
23:12:57.959 [DEBUG] [o.o.i.r.i.b.GeneralBroadcaster:46   ] - broadcaster 'org.openhab.io.rest.internal.broadcaster.GeneralBroadcaster$1@1b38868' is empty
23:12:58.051 [DEBUG] [.r.i.resources.SitemapResource:146  ] - Received HTTP GET request at 'sitemaps/default/default' for media type 'null'.
23:13:08.606 [DEBUG] [o.i.r.i.resources.ItemResource:205  ] - Received HTTP POST request at 'items/Light_Salon_Dimm' with value '48'.
23:13:08.611 [INFO ] [runtime.busevents             :22   ] - Light_Salon_Dimm received command 48
23:13:08.615 [DEBUG] [bticino.internal.BticinoDevice:134  ] - Gateway [default], Command '48' received for item Light_Salon_Dimm
23:13:08.618 [INFO ] [aminck.openwebnet.ProtocolRead:38   ] - Instance created for message [1*32]
23:13:08.622 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : who, Value : 1
23:13:08.626 [INFO ] [aminck.openwebnet.ProtocolRead:43   ] - addProperty Key : address, Value : 32
23:13:08.629 [WARN ] [bticino.internal.BticinoDevice:177  ] - Received unknown command type for lighting: 'org.openhab.core.library.types.PercentType'
23:13:08.633 [ERROR] [vlaminck.openwebnet.OpenWebNet:94   ] - onCommand error : null
23:13:08.995 [DEBUG] [o.o.i.r.i.b.GeneralBroadcaster:46   ] - broadcaster 'org.openhab.io.rest.internal.broadcaster.GeneralBroadcaster$1@1b38868' is empty
23:13:10.119 [DEBUG] [.r.i.resources.SitemapResource:146  ] - Received HTTP GET request at 'sitemaps/default/default' for media type 'null'.
```

Btw, When i use my Android Apps, the percentage was affiched and they wasn't when i use Increase or Decrease (but they works to dim my light ^^ )

Thanks for helping.
e4

--- a/bundles/binding/org.openhab.binding.bticino/src/main/java/be/devlaminck/openwebnet/MonitorSessionThread.java
+++ b/bundles/binding/org.openhab.binding.bticino/src/main/java/be/devlaminck/openwebnet/MonitorSessionThread.java
@@ -17,7 +17,7 @@ import org.slf4j.LoggerFactory;
 import com.myhome.fcrisciani.connector.MyHomeJavaConnector;
 
 /**
- * This thread reads every event on the bticino bus and the converts & publishes
+ * This thread reads every event on the bticino bus and then converts & publishes
  * it to the openhab
  *
  * Based on code from Mauro Cicolella (as part of the FREEDOMOTIC framework)
@@ -40,13 +40,16 @@ public class MonitorSessionThread extends Thread {
     @Override
     public void run() {
         // connect to own gateway
+        logger.debug("Connecting to ipaddress {} on port {}", ipAddress, port);
         pluginReference.myPlant = new MyHomeJavaConnector(ipAddress, port);
         try {
             pluginReference.myPlant.startMonitoring();
             while (!Thread.interrupted()) {
                 try {
                     String readFrame = pluginReference.myPlant.readMonitoring();
-                    buildEventFromFrame(readFrame);
+                    if (readFrame != null) {
+                        buildEventFromFrame(readFrame);
+                    }
                 } catch (InterruptedException ex) {
                     logger.error("MonitorSessionThread.run, exception : " + ex.getMessage());
                 }

--- a/bundles/binding/org.openhab.binding.bticino/src/main/java/be/devlaminck/openwebnet/OpenWebNet.java
+++ b/bundles/binding/org.openhab.binding.bticino/src/main/java/be/devlaminck/openwebnet/OpenWebNet.java
@@ -49,7 +49,7 @@ public class OpenWebNet extends Thread {
     /*
      * OWN Diagnostic Frames
      */
-    private final static String LIGHTNING_DIAGNOSTIC_FRAME = "*#1*0##";
+    private final static String LIGHTING_DIAGNOSTIC_FRAME = "*#1*0##";
     private final static String AUTOMATIONS_DIAGNOSTIC_FRAME = "*#2*0##";
     private final static String ALARM_DIAGNOSTIC_FRAME = "*#5##";
     private final static String POWER_MANAGEMENT_DIAGNOSTIC_FRAME = "*#3##";
@@ -127,8 +127,8 @@ public class OpenWebNet extends Thread {
     // sends diagnostic frames to initialize the system
     public void initSystem() {
         try {
-            logger.info("Sending " + LIGHTNING_DIAGNOSTIC_FRAME + " frame to (re)initialize LIGHTNING");
-            myPlant.sendCommandSync(LIGHTNING_DIAGNOSTIC_FRAME);
+            logger.info("Sending " + LIGHTING_DIAGNOSTIC_FRAME + " frame to (re)initialize LIGHTING");
+            myPlant.sendCommandSync(LIGHTING_DIAGNOSTIC_FRAME);
             logger.info("Sending " + AUTOMATIONS_DIAGNOSTIC_FRAME + " frame to (re)initialize AUTOMATIONS");
             myPlant.sendCommandSync(AUTOMATIONS_DIAGNOSTIC_FRAME);
             logger.info("Sending " + ALARM_DIAGNOSTIC_FRAME + " frame to (re)initialize ALARM");

--- a/bundles/binding/org.openhab.binding.bticino/src/main/java/com/myhome/fcrisciani/connector/MyHomeJavaConnector.java
+++ b/bundles/binding/org.openhab.binding.bticino/src/main/java/com/myhome/fcrisciani/connector/MyHomeJavaConnector.java
@@ -16,6 +16,9 @@ import java.net.Socket;
 import java.util.ArrayList;
 import java.util.concurrent.Semaphore;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.myhome.fcrisciani.datastructure.action.Action;
 import com.myhome.fcrisciani.datastructure.command.CommandOPEN;
 import com.myhome.fcrisciani.datastructure.command.DelayInterval;
@@ -36,6 +39,9 @@ import com.myhome.fcrisciani.queue.PriorityQueueThread;
  * @since 1.7.0
  */
 public class MyHomeJavaConnector {
+
+    private static final Logger logger = LoggerFactory.getLogger(MyHomeJavaConnector.class);
+
     // ----- TYPES ----- //
 
     // ---- MEMBERS ---- //
@@ -100,7 +106,7 @@ public class MyHomeJavaConnector {
     }
 
     /**
-     * Receive message form a monitor socket
+     * Receive message from a monitor socket
      * 
      * @param sk
      *            monitor socket used to read commands
@@ -129,6 +135,7 @@ public class MyHomeJavaConnector {
         super();
         this.ip = ip;
         this.port = port;
+        logger.debug("Created MyHomeJavaConnector with ip = {} and port = {}", ip, port);
         this.commandMutex = new Semaphore(1, true);
         this.commandQueue = new PriorityCommandQueue();
         this.commandQueueThread = new Thread(new PriorityQueueThread(this, commandQueue), "TailThread");
@@ -319,12 +326,11 @@ public class MyHomeJavaConnector {
                 }
                 retry++;
                 Thread.sleep(1000);
-                System.err.println("Monitor connection problem retry temptative: " + retry);
+                logger.error("Monitor connection problem. Attempting retry {}.", retry);
                 try {
                     startMonitoring();
                 } catch (IOException e1) {
                 }
-                continue;
             }
         } while (result == null);
 

--- a/bundles/binding/org.openhab.binding.bticino/src/main/java/org/openhab/binding/bticino/internal/BticinoDevice.java
+++ b/bundles/binding/org.openhab.binding.bticino/src/main/java/org/openhab/binding/bticino/internal/BticinoDevice.java
@@ -18,9 +18,13 @@ import org.openhab.core.library.items.RollershutterItem;
 import org.openhab.core.library.items.SwitchItem;
 import org.openhab.core.library.types.DecimalType;
 import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.PercentType;
+import org.openhab.core.library.types.IncreaseDecreaseType;
 import org.openhab.core.library.types.StopMoveType;
 import org.openhab.core.library.types.UpDownType;
 import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -29,9 +33,9 @@ import be.devlaminck.openwebnet.OpenWebNet;
 import be.devlaminck.openwebnet.ProtocolRead;
 
 /**
- * This class connects to the openweb gateway of bticino (MH200(N)) It opens a
- * monitor session to retrieve the events And creates (for every) command
- * received a command on the bus
+ * This class connects to the openweb gateway of bticino (MH200(N)). It opens a
+ * monitor session to retrieve the events and creates, for every command
+ * received, a command on the bus.
  *
  * @author Tom De Vlaminck
  * @serial 1.0
@@ -125,7 +129,7 @@ public class BticinoDevice implements IBticinoEventListener {
     public void receiveCommand(String itemName, Command command, BticinoBindingConfig itemBindingConfig) {
         try {
             synchronized (m_lock) {
-                // An command is received from the openHab system
+                // A command is received from the openHab system;
                 // analyse it and execute it
                 logger.debug("Gateway [" + m_gateway_id + "], Command '{}' received for item {}",
                         (Object[]) new String[] { command.toString(), itemName });
@@ -138,10 +142,32 @@ public class BticinoDevice implements IBticinoEventListener {
                 switch (l_who) {
                     // Lights
                     case 1: {
-                        if (OnOffType.ON.equals(command)) {
-                            l_pr.addProperty("what", "1");
-                        } else {
-                            l_pr.addProperty("what", "0");
+                        if (command instanceof IncreaseDecreaseType ) {
+                            if( IncreaseDecreaseType.INCREASE.equals(command) ) {
+                                logger.debug("Light received INCREASE command.");
+                                l_pr.addProperty("what", "30");
+                            }
+                            else {
+                                logger.debug("Light received DECREASE command.");
+                                l_pr.addProperty("what", "31");
+                            }
+                        }
+                        else if (command instanceof PercentType) {
+                            PercentType pType = (PercentType)command;
+                            //take percentage and divide by 10, round 1 (ie 0 to 10 is the result, nothing else)
+                            int percentValue = (int)(Math.floor(pType.intValue()/10F));
+                            logger.debug("Set light value to {}", percentValue);
+                            l_pr.addProperty("what", String.valueOf(percentValue));
+                        }
+                        else if (command instanceof OnOffType) {
+                            if (OnOffType.ON.equals(command)) {
+                                l_pr.addProperty("what", "1");
+                            } else {
+                                l_pr.addProperty("what", "0");
+                            }
+                        }
+                        else {
+                            logger.warn("Received unknown command type for lighting: '{}'", command.getClass().getName());
                         }
                         break;
                     }

